File: src_py/pipeline.py
import json
import os
from .validation import validate_batch
from .normalization import normalize_event
from .deduplication import cluster_and_deduplicate
from .attribution import determine_attribution
from .reconciliation import reconcile_state

EVENT_TYPE_ORDER = {
    'page_visit': 1,
    'campaign_click': 2,
    'form_submission': 3,
    'email_open': 4,
    'email_click': 5,
    'contacted': 6,
    'qualified': 7,
    'converted': 8,
    'lost': 9,
}

def execute_pipeline(raw_events: list, output_dir="output", write_to_disk=False):
    # Deduplicate event_id
    seen_ids = set()
    dedup_raw = []
    for r in raw_events:
        eid = str(r.get('event_id') or '').strip()
        if eid:
            if eid in seen_ids:
                continue
            seen_ids.add(eid)
        dedup_raw.append(r)

    valid_raw, val_errors = validate_batch(dedup_raw)
    normalized = [normalize_event(r, i) for i, r in enumerate(valid_raw)]
    clusters, dup_audit = cluster_and_deduplicate(normalized)

    all_audit = list(dup_audit)
    canonical_leads = []

    total_conflicts = 0
    total_qualified = 0
    total_converted = 0

    for cluster in clusters:
        # Sort journey
        events = sorted(cluster['events'], key=lambda e: (e['timestamp'], EVENT_TYPE_ORDER.get(e['event_type'], 50), e['event_id']))
        timeline = [{
            "event_id": e['event_id'],
            "event_type": e['event_type'],
            "timestamp": e['timestamp'],
            "campaign": e['campaign'],
            "source": e['source'],
            "status": e['status']
        } for e in events]

        attr_exp, attr_audit = determine_attribution(cluster['canonical_id'], timeline)
        all_audit.extend(attr_audit)

        curr_state, has_conf, confs, state_audit = reconcile_state(cluster['canonical_id'], timeline)
        all_audit.extend(state_audit)

        if has_conf:
            total_conflicts += 1
        if curr_state in ('Qualified', 'Converted'):
            total_qualified += 1
        if curr_state == 'Converted':
            total_converted += 1

        canonical_leads.append({
            "lead_id": cluster['canonical_id'],
            "canonical_email": cluster['primary_email'],
            "phone": cluster['primary_phone'],
            "name": cluster['primary_name'],
            "current_state": curr_state,
            "attributed_campaign": attr_exp['selected_campaign'],
            "attributed_source": attr_exp['selected_source'],
            "attribution_explanation": attr_exp,
            "duplicate_count": cluster['duplicate_count'],
            "interaction_count": len(timeline),
            "has_conflict": has_conf,
            "conflicts": confs,
            "timeline": timeline
        })

    canonical_leads.sort(key=lambda l: l['lead_id'])
    all_audit.sort(key=lambda a: (a['timestamp'], a['lead_id'], a['id']))

    summary = {
        "total_raw_events": len(raw_events),
        "valid_events": len(valid_raw),
        "rejected_records": len(val_errors),
        "unique_leads": len(canonical_leads),
        "qualified_leads": total_qualified,
        "converted_leads": total_converted,
        "reconciliation_conflicts": total_conflicts
    }

    if write_to_disk:
        os.makedirs(output_dir, exist_ok=True)
        with open(os.path.join(output_dir, 'reconciled_leads.json'), 'w') as f:
            json.dump(canonical_leads, f, indent=2)

    return {
        "summary": summary,
        "leads": canonical_leads,
        "audit_trail": all_audit,
        "validation_errors": val_errors
    }

File: src/engine/pipeline.ts
import {
  RawEvent,
  CanonicalLead,
  PipelineResult,
  PipelineSummary,
  AuditRecord,
  ValidationError,
  SEOKeywordRow,
} from '../types.js';
import { validateEventBatch } from './validation.js';
import { normalizeEventBatch } from './normalization.js';
import { clusterAndDeduplicateEvents } from './deduplication.js';
import { reconstructLeadJourney } from './journey.js';
import { determineCampaignAttribution } from './attribution.js';
import { reconcileLeadState } from './reconciliation.js';
import { calculateLeadScore } from './scoring.js';
import {
  generateCampaignReport,
  generateSourceReport,
  exportCampaignReportCSV,
  exportLeadsCSV,
  exportAuditCSV,
} from './reporting.js';
import { processSEOKeywords, exportSEOReportCSV } from './seo.js';

export interface PipelineOptions {
  customSourcePrecedence?: string[];
  outputDir?: string;
  writeToDisk?: boolean;
  seoData?: any[];
}

export function executeReconciliationPipeline(
  rawEvents: RawEvent[],
  options: PipelineOptions = {}
): PipelineResult {
  const startTime = Date.now();

  // --- Step 0: Event ID Deduplication & Idempotency Key Handling ---
  // If the exact same event_id appears multiple times in raw input, deduplicate it deterministically
  const seenEventIds = new Set<string>();
  const deduplicatedRawEvents: RawEvent[] = [];
  let duplicateEventIdCount = 0;

  for (const raw of rawEvents) {
    const rawId = raw && raw.event_id ? String(raw.event_id).trim() : null;
    if (rawId) {
      if (seenEventIds.has(rawId)) {
        duplicateEventIdCount++;
        continue; // Idempotent skip of duplicate raw event_id
      }
      seenEventIds.add(rawId);
    }
    deduplicatedRawEvents.push(raw);
  }

  // --- Step 1: Validation ---
  const { validRawEvents, validationErrors } = validateEventBatch(deduplicatedRawEvents);

  // --- Step 2: Normalization ---
  const normalizedEvents = normalizeEventBatch(validRawEvents);

  // --- Step 3: Deduplication & Lead Clustering ---
  const { clusters, duplicateAuditRecords } = clusterAndDeduplicateEvents(normalizedEvents);

  const allAuditRecords: AuditRecord[] = [...duplicateAuditRecords];
  const canonicalLeads: CanonicalLead[] = [];

  let totalConflictsCount = 0;
  let totalQualifiedCount = 0;
  let totalConvertedCount = 0;

  // --- Step 4: Journey Reconstruction, Attribution, Reconciliation, Scoring for each Lead ---
  for (const cluster of clusters) {
    // a. Journey Reconstruction
    const { timeline, lateEventAuditRecords } = reconstructLeadJourney(
      cluster.canonical_id,
      cluster.events
    );
    allAuditRecords.push(...lateEventAuditRecords);

    // b. Campaign Attribution
    const { attribution, attributionAuditRecords } = determineCampaignAttribution(
      cluster.canonical_id,
      timeline,
      options.customSourcePrecedence
    );
    allAuditRecords.push(...attributionAuditRecords);

    // c. State Reconciliation & Conflict Detection
    const { currentState, hasConflict, conflicts, stateAuditRecords } = reconcileLeadState(
      cluster.canonical_id,
      timeline
    );
    allAuditRecords.push(...stateAuditRecords);

    if (hasConflict) totalConflictsCount++;
    if (currentState === 'Qualified' || currentState === 'Converted') totalQualifiedCount++;
    if (currentState === 'Converted') totalConvertedCount++;

    // d. Lead Quality Scoring
    const score = calculateLeadScore(timeline, currentState, cluster.primary_phone);

    const firstSeen = timeline.length > 0 ? timeline[0].timestamp : new Date().toISOString();
    const lastSeen = timeline.length > 0 ? timeline[timeline.length - 1].timestamp : firstSeen;

    canonicalLeads.push({
      lead_id: cluster.canonical_id,
      canonical_email: cluster.primary_email,
      phone: cluster.primary_phone,
      name: cluster.primary_name,
      current_state: currentState,
      attributed_campaign: attribution.selected_campaign,
      attributed_source: attribution.selected_source,
      attribution_explanation: attribution,
      duplicate_count: cluster.duplicate_count,
      interaction_count: timeline.length,
      has_conflict: hasConflict,
      conflicts,
      timeline,
      merged_event_ids: cluster.events.map(e => e.event_id),
      score,
      first_seen: firstSeen,
      last_seen: lastSeen,
    });
  }

  // Sort canonical leads deterministically by lead_id
  canonicalLeads.sort((a, b) => a.lead_id.localeCompare(b.lead_id));

  // Sort audit records deterministically by timestamp, then lead_id, then id
  allAuditRecords.sort((a, b) => {
    const timeDiff = a.timestamp.localeCompare(b.timestamp);
    if (timeDiff !== 0) return timeDiff;
    const leadDiff = a.lead_id.localeCompare(b.lead_id);
    if (leadDiff !== 0) return leadDiff;
    return a.id.localeCompare(b.id);
  });

  // --- Step 5: Campaign & Source Reporting ---
  const campaignReport = generateCampaignReport(canonicalLeads);
  const sourceReport = generateSourceReport(canonicalLeads);

  // Optional SEO report
  let seoReport: SEOKeywordRow[] | undefined;
  if (options.seoData && options.seoData.length > 0) {
    seoReport = processSEOKeywords(options.seoData);
  }

  const executionTimeMs = Date.now() - startTime;
  const overallConversionRate =
    canonicalLeads.length > 0
      ? Number(((totalConvertedCount / canonicalLeads.length) * 100).toFixed(2))
      : 0.0;

  const summary: PipelineSummary = {
    total_raw_events: rawEvents.length,
    valid_events: validRawEvents.length,
    rejected_records: validationErrors.length,
    duplicate_records:
      canonicalLeads.reduce((sum, l) => sum + l.duplicate_count, 0) + duplicateEventIdCount,
    unique_leads: canonicalLeads.length,
    qualified_leads: totalQualifiedCount,
    converted_leads: totalConvertedCount,
    overall_conversion_rate: overallConversionRate,
    reconciliation_conflicts: totalConflictsCount,
    processed_at: new Date().toISOString(),
    execution_time_ms: executionTimeMs,
  };

  const result: PipelineResult = {
    summary,
    leads: canonicalLeads,
    campaign_report: campaignReport,
    source_report: sourceReport,
    audit_trail: allAuditRecords,
    validation_errors: validationErrors,
    seo_report: seoReport,
  };

  return result;
}
